File: tools/build_google_routes_cache.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(optparse)
  library(jsonlite)
})

option_list <- list(
  make_option(c("--flows_csv"), type = "character", default = "data/derived/faf_top_od_flows.csv"),
  make_option(c("--zones_csv"), type = "character", default = "data/derived/faf_zone_centroids.csv"),
  make_option(c("--out_cache_csv"), type = "character", default = "data/derived/google_routes_od_cache.csv"),
  make_option(c("--out_dist_csv"), type = "character", default = "data/derived/google_routes_distance_distributions.csv"),
  make_option(c("--out_meta_json"), type = "character", default = "data/derived/google_routes_metadata.json"),
  make_option(c("--api_key"), type = "character", default = ""),
  make_option(c("--max_pairs"), type = "integer", default = 400L),
  make_option(c("--sleep_ms"), type = "integer", default = 0L),
  make_option(c("--dry_run"), action = "store_true", default = FALSE)
)

opt <- parse_args(OptionParser(
  usage = "%prog [options]",
  description = "Build cached road-distance OD table and simulation distance distributions using Google Routes API.",
  option_list = option_list
))

log_info <- function(...) message("[google_routes] ", paste0(..., collapse = ""))

weighted_quantile <- function(x, w, probs) {
  o <- order(x)
  x <- x[o]; w <- w[o]
  cw <- cumsum(w) / sum(w)
  vapply(probs, function(p) x[which(cw >= p)[1]], numeric(1))
}

distance_id_for_scenario <- function(s) {
  s <- toupper(trimws(s))
  if (s == "CENTRALIZED") return("dist_centralized_food_truck_2024")
  if (s == "REGIONALIZED") return("dist_regionalized_food_truck_2024")
  paste0("dist_", tolower(gsub("[^A-Za-z0-9]+", "_", s)), "_google_routes")
}

empty_dist_schema <- function() {
  data.frame(
    distance_distribution_id = character(),
    scenario_id = character(),
    source_zip = character(),
    commodity_filter = character(),
    mode_filter = character(),
    distance_model = character(),
    p05_miles = numeric(),
    p50_miles = numeric(),
    p95_miles = numeric(),
    mean_miles = numeric(),
    min_miles = numeric(),
    max_miles = numeric(),
    n_records = integer(),
    status = character(),
    source_id = character(),
    notes = character(),
    stringsAsFactors = FALSE
  )
}

get_api_key <- function() {
  key <- trimws(opt$api_key)
  if (nzchar(key)) return(key)
  key <- Sys.getenv("GOOGLE_MAPS_API_KEY", unset = "")
  if (nzchar(key)) return(trimws(key))
  stop("Missing API key. Provide --api_key or set GOOGLE_MAPS_API_KEY.")
}

call_route <- function(lat1, lon1, lat2, lon2, api_key) {
  body <- toJSON(list(
    origin = list(location = list(latLng = list(latitude = lat1, longitude = lon1))),
    destination = list(location = list(latLng = list(latitude = lat2, longitude = lon2))),
    travelMode = "DRIVE",
    routingPreference = "TRAFFIC_UNAWARE",
    units = "IMPERIAL"
  ), auto_unbox = TRUE)
  tf <- tempfile(fileext = ".json")
  writeLines(body, tf)
  on.exit(unlink(tf), add = TRUE)

  args <- c(
    "-sS",
    "-X", "POST",
    "https://routes.googleapis.com/directions/v2:computeRoutes",
    "-H", "Content-Type: application/json",
    "-H", paste0("X-Goog-Api-Key: ", api_key),
    "-H", "X-Goog-FieldMask: routes.distanceMeters,routes.duration",
    "--data-binary", paste0("@", tf)
  )
  out <- system2("curl", args, stdout = TRUE, stderr = TRUE)
  status <- attr(out, "status")
  if (!is.null(status) && status != 0) {
    return(list(ok = FALSE, error = paste(out, collapse = "\n")))
  }
  txt <- paste(out, collapse = "\n")
  obj <- tryCatch(fromJSON(txt, simplifyVector = FALSE), error = function(e) NULL)
  if (is.null(obj) || is.null(obj$routes) || length(obj$routes) == 0) {
    return(list(ok = FALSE, error = "No routes returned"))
  }
  r0 <- obj$routes[[1]]
  dm <- suppressWarnings(as.numeric(r0$distanceMeters))
  dur <- suppressWarnings(as.numeric(gsub("s$", "", as.character(r0$duration))))
  if (!is.finite(dm)) return(list(ok = FALSE, error = "Missing distanceMeters"))
  list(ok = TRUE, miles = dm / 1609.344, minutes = if (is.finite(dur)) dur / 60 else NA_real_)
}

if (!file.exists(opt$flows_csv)) stop("Flows CSV not found: ", opt$flows_csv)
if (!file.exists(opt$zones_csv)) stop("Zones CSV not found: ", opt$zones_csv)

flows <- utils::read.csv(opt$flows_csv, stringsAsFactors = FALSE)
zones <- utils::read.csv(opt$zones_csv, stringsAsFactors = FALSE)
for (nm in c("origin_id", "dest_id", "scenario_id", "tons")) {
  if (!(nm %in% names(flows))) stop("flows_csv missing required column: ", nm)
}
for (nm in c("zone_id", "lat", "lon")) {
  if (!(nm %in% names(zones))) stop("zones_csv missing required column: ", nm)
}
zones$zone_id <- as.character(zones$zone_id)
zones$lat <- suppressWarnings(as.numeric(zones$lat))
zones$lon <- suppressWarnings(as.numeric(zones$lon))

od <- unique(flows[, c("origin_id", "dest_id"), drop = FALSE])
if (nrow(od) > opt$max_pairs) od <- od[seq_len(opt$max_pairs), , drop = FALSE]

od$origin_id <- as.character(od$origin_id)
od$dest_id <- as.character(od$dest_id)
o <- zones[match(od$origin_id, zones$zone_id), c("lat", "lon")]
d <- zones[match(od$dest_id, zones$zone_id), c("lat", "lon")]
names(o) <- c("origin_lat", "origin_lon")
names(d) <- c("dest_lat", "dest_lon")
od <- cbind(od, o, d)
od <- od[is.finite(od$origin_lat) & is.finite(od$origin_lon) & is.finite(od$dest_lat) & is.finite(od$dest_lon), , drop = FALSE]

if (nrow(od) == 0) stop("No valid OD pairs after joining zone centroids.")

api_key <- if (isTRUE(opt$dry_run)) "" else get_api_key()
out <- vector("list", nrow(od))
for (i in seq_len(nrow(od))) {
  row <- od[i, , drop = FALSE]
  if (isTRUE(opt$dry_run)) {
    out[[i]] <- data.frame(
      origin_id = row$origin_id, dest_id = row$dest_id,
      road_distance_miles = NA_real_, road_duration_minutes = NA_real_,
      status = "DRY_RUN", error = "", stringsAsFactors = FALSE
    )
    next
  }
  res <- call_route(row$origin_lat, row$origin_lon, row$dest_lat, row$dest_lon, api_key)
  out[[i]] <- data.frame(
    origin_id = row$origin_id, dest_id = row$dest_id,
    road_distance_miles = if (isTRUE(res$ok)) res$miles else NA_real_,
    road_duration_minutes = if (isTRUE(res$ok)) res$minutes else NA_real_,
    status = if (isTRUE(res$ok)) "OK" else "ERROR",
    error = if (isTRUE(res$ok)) "" else as.character(res$error),
    stringsAsFactors = FALSE
  )
  if (opt$sleep_ms > 0) Sys.sleep(opt$sleep_ms / 1000)
}
cache <- do.call(rbind, out)
cache$generated_at_utc <- format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC")
cache$api_provider <- "google_routes_v2"

joined <- merge(
  flows[, c("origin_id", "dest_id", "scenario_id", "tons"), drop = FALSE],
  cache[, c("origin_id", "dest_id", "road_distance_miles", "status"), drop = FALSE],
  by = c("origin_id", "dest_id"),
  all.x = FALSE
)
joined$tons <- suppressWarnings(as.numeric(joined$tons))
joined$road_distance_miles <- suppressWarnings(as.numeric(joined$road_distance_miles))
joined <- joined[is.finite(joined$tons) & joined$tons > 0 & is.finite(joined$road_distance_miles), , drop = FALSE]

dist_rows <- list()
for (s in unique(as.character(flows$scenario_id))) {
  x <- joined[joined$scenario_id == s, , drop = FALSE]
  if (nrow(x) == 0) next
  q <- weighted_quantile(x$road_distance_miles, x$tons, probs = c(0.05, 0.5, 0.95))
  mn <- sum(x$road_distance_miles * x$tons) / sum(x$tons)
  dist_rows[[length(dist_rows) + 1]] <- data.frame(
    distance_distribution_id = distance_id_for_scenario(s),
    scenario_id = s,
    source_zip = "google_routes_api_cached_od",
    commodity_filter = "food_sctg_01_08",
    mode_filter = "truck",
    distance_model = "triangular_fit",
    p05_miles = q[[1]], p50_miles = q[[2]], p95_miles = q[[3]],
    mean_miles = mn,
    min_miles = min(x$road_distance_miles, na.rm = TRUE),
    max_miles = max(x$road_distance_miles, na.rm = TRUE),
    n_records = nrow(x),
    status = "OK",
    source_id = "google_routes_api_cached_od",
    notes = "Weighted by tons from faf_top_od_flows.csv and Google Routes API cached OD distances.",
    stringsAsFactors = FALSE
  )
}
dist_df <- if (length(dist_rows) > 0) do.call(rbind, dist_rows) else empty_dist_schema()

dir.create(dirname(opt$out_cache_csv), recursive = TRUE, showWarnings = FALSE)
utils::write.csv(cache, opt$out_cache_csv, row.names = FALSE)
utils::write.csv(dist_df, opt$out_dist_csv, row.names = FALSE)

meta <- list(
  generated_at_utc = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
  api_provider = "google_routes_v2",
  dry_run = isTRUE(opt$dry_run),
  pairs_requested = nrow(od),
  pairs_ok = sum(cache$status == "OK"),
  pairs_error = sum(cache$status != "OK")
)
writeLines(toJSON(meta, auto_unbox = TRUE, pretty = TRUE), opt$out_meta_json)

log_info("Wrote ", opt$out_cache_csv)
log_info("Wrote ", opt$out_dist_csv)
log_info("Wrote ", opt$out_meta_json)

File: tools/build_google_routes_cache.py
#!/usr/bin/env python3

"""
build_google_routes_cache.py
This script replicates the functionality of the R build_google_routes_cache.R script
using Python and the requests HTTP client. It reads origin–destination flows and
zone centroid coordinates, calls the Google Routes API to compute road distances
and durations, writes a cached OD table, computes weighted distance distributions
by scenario, and outputs metadata. Authentication can be via OAuth (using gcloud)
and/or API key.

Usage example:

    python build_google_routes_cache.py \
        --flows_csv data/derived/faf_top_od_flows.csv \
        --zones_csv data/derived/faf_zone_centroids.csv \
        --out_cache_csv data/derived/google_routes_od_cache.csv \
        --out_dist_csv data/derived/google_routes_distance_distributions.csv \
        --out_meta_json data/derived/google_routes_metadata.json \
        --max_pairs 5 \
        --auth_mode oauth \
        --user_project coldchain-freight-ttp211

Note: To use OAuth, gcloud must be installed and configured. An API key can be
provided via --api_key or the GOOGLE_MAPS_API_KEY environment variable. When
using OAuth, the script also sends the API key if available.
"""

import argparse
import csv
import json
import os
import subprocess
import sys
import time
from datetime import datetime, timezone
from typing import List, Tuple, Dict, Any

try:
    import requests
except ImportError:
    sys.stderr.write("The 'requests' library is required. Install it via pip install requests\n")
    raise


def weighted_quantile(values: List[float], weights: List[float], probs: List[float]) -> List[float]:
    """Compute weighted quantiles. values and weights must be the same length."""
    # Sort by value
    sorted_data = sorted(zip(values, weights), key=lambda x: x[0])
    xs, ws = zip(*sorted_data)
    total_weight = sum(ws)
    cum_weights = []
    csum = 0.0
    for w in ws:
        csum += w
        cum_weights.append(csum / total_weight)
    quantiles = []
    for p in probs:
        # Find first index where cumulative weight >= p
        for idx, cw in enumerate(cum_weights):
            if cw >= p:
                quantiles.append(xs[idx])
                break
        else:
            quantiles.append(xs[-1])
    return quantiles


def distance_id_for_scenario(s: str) -> str:
    s = s.strip().upper()
    if s == "CENTRALIZED":
        return "dist_centralized_food_truck_2024"
    if s == "REGIONALIZED":
        return "dist_regionalized_food_truck_2024"
    import re
    return "dist_" + re.sub(r"[^A-Za-z0-9]+", "_", s.lower()) + "_google_routes"


def get_api_key(cli_key: str) -> str:
    if cli_key:
        return cli_key.strip()
    env_key = os.environ.get("GOOGLE_MAPS_API_KEY", "").strip()
    return env_key


def run_gcloud_command(args: List[str]) -> Tuple[bool, str, str]:
    """Run a gcloud command and return (success, stdout, detail)."""
    try:
        result = subprocess.run([
            "gcloud", *args
        ], stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, check=False)
    except FileNotFoundError:
        return False, "", "gcloud not found"
    success = (result.returncode == 0)
    stdout = result.stdout.strip().split("\n")
    value = stdout[-1].strip() if stdout else ""
    if value == "(unset)":
        value = ""
    detail = result.stdout + result.stderr
    return success, value, detail


def get_oauth_access_token() -> Tuple[bool, str, str]:
    """Try to obtain an OAuth access token via gcloud."""
    for cmd in [
        ["auth", "application-default", "print-access-token"],
        ["auth", "print-access-token"]
    ]:
        ok, token, detail = run_gcloud_command(cmd)
        if ok and token:
            return True, token, ""
    # failure
    return False, "", detail


def get_user_project(cli_project: str) -> Tuple[bool, str, str]:
    if cli_project:
        return True, cli_project.strip(), ""
    env_project = os.environ.get("GOOGLE_CLOUD_PROJECT") or os.environ.get("GCLOUD_PROJECT")
    if env_project and env_project.strip():
        return True, env_project.strip(), ""
    ok, value, detail = run_gcloud_command(["config", "get-value", "project", "--quiet"])
    if ok and value:
        return True, value.strip(), ""
    return False, "", detail


def resolve_auth(auth_mode: str, cli_api_key: str, cli_user_project: str) -> Dict[str, Any]:
    mode = (auth_mode or "auto").lower().strip()
    api_key = get_api_key(cli_api_key)
    if mode == "api_key":
        if not api_key:
            raise RuntimeError("Missing API key; provide --api_key or set GOOGLE_MAPS_API_KEY")
        return {
            "mode_requested": mode,
            "mode_used": "api_key",
            "auth_header": [f"X-Goog-Api-Key: {api_key}"],
            "user_project": ""
        }
    if mode == "oauth":
        ok_token, token, detail = get_oauth_access_token()
        if not ok_token:
            raise RuntimeError(f"Unable to obtain OAuth token: {detail}")
        ok_proj, proj, proj_detail = get_user_project(cli_user_project)
        if not ok_proj:
            raise RuntimeError(f"Unable to determine user project: {proj_detail}")
        headers = [f"Authorization: Bearer {token}"]
        # include API key if available
        if api_key:
            headers.append(f"X-Goog-Api-Key: {api_key}")
        return {
            "mode_requested": mode,
            "mode_used": "oauth",
            "auth_header": headers,
            "user_project": proj
        }
    # auto mode
    # try oauth first
    ok_token, token, _ = get_oauth_access_token()
    if ok_token and token:
        ok_proj, proj, _ = get_user_project(cli_user_project)
        if ok_proj:
            headers = [f"Authorization: Bearer {token}"]
            if api_key:
                headers.append(f"X-Goog-Api-Key: {api_key}")
            return {
                "mode_requested": mode,
                "mode_used": "oauth",
                "auth_header": headers,
                "user_project": proj
            }
    # fallback to api key
    if api_key:
        return {
            "mode_requested": mode,
            "mode_used": "api_key",
            "auth_header": [f"X-Goog-Api-Key: {api_key}"],
            "user_project": ""
        }
    raise RuntimeError("No valid authentication method available; set OAuth credentials or provide API key")


def call_route(lat1: float, lon1: float, lat2: float, lon2: float, headers: List[str], user_project: str) -> Dict[str, Any]:
    """Call the Google Routes API for a single origin–destination pair."""
    url = "https://routes.googleapis.com/directions/v2:computeRoutes"
    body = {
        "origin": {
            "location": {
                "latLng": {
                    "latitude": lat1,
                    "longitude": lon1
                }
            }
        },
        "destination": {
            "location": {
                "latLng": {
                    "latitude": lat2,
                    "longitude": lon2
                }
            }
        },
        "travelMode": "DRIVE",
        "routingPreference": "TRAFFIC_UNAWARE",
        "units": "IMPERIAL"
    }
    req_headers = {
        "Content-Type": "application/json",
        "X-Goog-FieldMask": "routes.distanceMeters,routes.duration"
    }
    if user_project:
        req_headers["X-Goog-User-Project"] = user_project
    for h in headers:
        if h.startswith("Authorization:"):
            val = h.split(":", 1)[1].strip()
            req_headers["Authorization"] = val
        elif h.startswith("X-Goog-Api-Key:"):
            val = h.split(":", 1)[1].strip()
            req_headers["X-Goog-Api-Key"] = val
    try:
        resp = requests.post(url, headers=req_headers, json=body)
    except Exception as e:
        return {"ok": False, "error": str(e), "miles": None, "minutes": None}
    if resp.status_code != 200:
        try:
            err = resp.text
        except Exception:
            err = f"HTTP {resp.status_code}"
        return {"ok": False, "error": err, "miles": None, "minutes": None}
    try:
        data = resp.json()
    except Exception:
        return {"ok": False, "error": "Failed to parse JSON", "miles": None, "minutes": None}
    routes = data.get("routes")
    if not routes:
        return {"ok": False, "error": "No routes returned", "miles": None, "minutes": None}
    r0 = routes[0]
    dm = r0.get("distanceMeters")
    dur = r0.get("duration")
    if dm is None:
        return {"ok": False, "error": "Missing distanceMeters", "miles": None, "minutes": None}
    try:
        miles = dm / 1609.344
    except Exception:
        miles = None
    try:
        # duration string like "19846s"
        minutes = float(dur[:-1]) / 60 if dur else None
    except Exception:
        minutes = None
    return {"ok": True, "error": "", "miles": miles, "minutes": minutes}


def read_csv_as_dict_list(filepath: str) -> List[Dict[str, str]]:
    rows = []
    with open(filepath, newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    return rows


def main():
    parser = argparse.ArgumentParser(description="Build Google Routes cache using Python requests")
    parser.add_argument("--flows_csv", default="data/derived/faf_top_od_flows.csv")
    parser.add_argument("--zones_csv", default="data/derived/faf_zone_centroids.csv")
    parser.add_argument("--out_cache_csv", default="data/derived/google_routes_od_cache.csv")
    parser.add_argument("--out_dist_csv", default="data/derived/google_routes_distance_distributions.csv")
    parser.add_argument("--out_meta_json", default="data/derived/google_routes_metadata.json")
    parser.add_argument("--api_key", default="", help="Google Maps API key")
    parser.add_argument("--auth_mode", default="auto", help="Authentication mode: auto|oauth|api_key")
    parser.add_argument("--user_project", default="", help="Quota/billing project for OAuth")
    parser.add_argument("--max_pairs", type=int, default=400)
    parser.add_argument("--sleep_ms", type=int, default=0)
    parser.add_argument("--dry_run", action="store_true")
    args = parser.parse_args()

    # Read input CSVs
    if not os.path.exists(args.flows_csv):
        sys.exit(f"Flows CSV not found: {args.flows_csv}")
    if not os.path.exists(args.zones_csv):
        sys.exit(f"Zones CSV not found: {args.zones_csv}")
    flows = read_csv_as_dict_list(args.flows_csv)
    zones = read_csv_as_dict_list(args.zones_csv)
    # Build zone lookup
    zone_lookup = {}
    for z in zones:
        zone_id = str(z.get("zone_id"))
        try:
            lat = float(z.get("lat", "nan"))
            lon = float(z.get("lon", "nan"))
        except Exception:
            lat = float("nan"); lon = float("nan")
        zone_lookup[zone_id] = (lat, lon)
    # Extract OD pairs
    od_pairs = []
    seen = set()
    for row in flows:
        oid = str(row.get("origin_id"))
        did = str(row.get("dest_id"))
        if (oid, did) not in seen:
            seen.add((oid, did))
            od_pairs.append((oid, did))
        if len(od_pairs) >= args.max_pairs:
            break
    # Resolve auth
    auth = resolve_auth(args.auth_mode, args.api_key, args.user_project)
    # Process each pair
    cache_rows = []
    for (oid, did) in od_pairs:
        lat1, lon1 = zone_lookup.get(str(oid), (float("nan"), float("nan")))
        lat2, lon2 = zone_lookup.get(str(did), (float("nan"), float("nan")))
        if not (lat1 == lat1 and lon1 == lon1 and lat2 == lat2 and lon2 == lon2):
            status = "ERROR"
            cache_rows.append({
                "origin_id": oid, "dest_id": did,
                "road_distance_miles": "",
                "road_duration_minutes": "",
                "status": status,
                "error": "Invalid coordinates"
            })
            continue
        if args.dry_run:
            cache_rows.append({
                "origin_id": oid, "dest_id": did,
                "road_distance_miles": "",
                "road_duration_minutes": "",
                "status": "DRY_RUN",
                "error": ""
            })
            continue
        res = call_route(lat1, lon1, lat2, lon2, auth["auth_header"], auth["user_project"])
        cache_rows.append({
            "origin_id": oid,
            "dest_id": did,
            "road_distance_miles": f"{res['miles']:.6f}" if res["ok"] and res["miles"] is not None else "",
            "road_duration_minutes": f"{res['minutes']:.6f}" if res["ok"] and res["minutes"] is not None else "",
            "status": "OK" if res["ok"] else "ERROR",
            "error": res.get("error", "") if not res["ok"] else ""
        })
        if args.sleep_ms > 0:
            time.sleep(args.sleep_ms / 1000.0)
    # Write cache CSV
    os.makedirs(os.path.dirname(args.out_cache_csv), exist_ok=True)
    with open(args.out_cache_csv, "w", newline="", encoding="utf-8") as f:
        fieldnames = ["origin_id", "dest_id", "road_distance_miles", "road_duration_minutes", "status", "error", "generated_at_utc", "api_provider"]
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        now = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
        for row in cache_rows:
            row_out = row.copy()
            row_out["generated_at_utc"] = now
            row_out["api_provider"] = "google_routes_v2_py"
            writer.writerow(row_out)
    # Compute weighted distance distributions
    # Build a lookup of road distances by origin-dest to join with flows
    dist_lookup = {(r["origin_id"], r["dest_id"]): r for r in cache_rows if r["status"] == "OK" and r["road_distance_miles"]}
    joined = []
    for row in flows:
        oid, did = str(row.get("origin_id")), str(row.get("dest_id"))
        key = (oid, did)
        if key in dist_lookup:
            dist_val = float(dist_lookup[key]["road_distance_miles"])
            tons = float(row.get("tons", "nan"))
            if tons > 0 and dist_val == dist_val:
                joined.append({
                    "scenario_id": str(row.get("scenario_id")),
                    "road_distance_miles": dist_val,
                    "tons": tons
                })
    # Group by scenario and compute quantiles
    dist_rows = []
    scenarios = set([j["scenario_id"] for j in joined])
    for s in scenarios:
        vals = [j["road_distance_miles"] for j in joined if j["scenario_id"] == s]
        wts = [j["tons"] for j in joined if j["scenario_id"] == s]
        if not vals:
            continue
        qs = weighted_quantile(vals, wts, [0.05, 0.5, 0.95])
        mean_m = sum(v * w for v, w in zip(vals, wts)) / sum(wts)
        dist_rows.append({
            "distance_distribution_id": distance_id_for_scenario(s),
            "scenario_id": s,
            "source_zip": "google_routes_api_cached_od",
            "commodity_filter": "food_sctg_01_08",
            "mode_filter": "truck",
            "distance_model": "triangular_fit",
            "p05_miles": qs[0],
            "p50_miles": qs[1],
            "p95_miles": qs[2],
            "mean_miles": mean_m,
            "min_miles": min(vals),
            "max_miles": max(vals),
            "n_records": len(vals),
            "status": "OK",
            "source_id": "google_routes_api_cached_od",
            "notes": "Weighted by tons from flows and Google Routes API cached OD distances."
        })
    # Write distribution CSV
    with open(args.out_dist_csv, "w", newline="", encoding="utf-8") as f:
        fieldnames = [
            "distance_distribution_id", "scenario_id", "source_zip", "commodity_filter",
            "mode_filter", "distance_model", "p05_miles", "p50_miles", "p95_miles",
            "mean_miles", "min_miles", "max_miles", "n_records", "status", "source_id", "notes"
        ]
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for row in dist_rows:
            writer.writerow(row)
    # Write metadata JSON
    meta = {
        "generated_at_utc": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "api_provider": "google_routes_v2_py",
        "auth_mode_requested": args.auth_mode,
        "auth_mode_used": auth["mode_used"],
        "user_project": auth["user_project"],
        "dry_run": args.dry_run,
        "pairs_requested": len(od_pairs),
        "pairs_ok": sum(1 for r in cache_rows if r["status"] == "OK"),
        "pairs_error": sum(1 for r in cache_rows if r["status"] != "OK")
    }
    with open(args.out_meta_json, "w", encoding="utf-8") as f:
        json.dump(meta, f, indent=2)
    print(f"Wrote {args.out_cache_csv}")
    print(f"Wrote {args.out_dist_csv}")
    print(f"Wrote {args.out_meta_json}")


if __name__ == "__main__":
    main()